client.go

```diff
@@ -211,6 +264,242 @@ func (c *Client) monitorChannel(ctx context.Context) {
+				case <-ctx.Done():
+					return
+				case <-timer.C:
+				}
+			case badCertificate:
+				// todo: recreate server certificate
+				// state = reconnectStateSechanDisconnected
+
+				debug.Printf("Recreating server certificate is not implemented yet, reconnection aborted")
+				state = aborted
+			case sechanReconnected:
+				debug.Printf("Secure channel reconnected")
+				if err := c.repairSession(); err != nil {
+					debug.Printf("Repair sessions has failed, %s", err.Error())
+					state = aborted
+					continue
+				}
+				debug.Printf("Session repaired")
+
+				// Resume all subscriptions
+				for _, sub := range c.subscriptions {
+					sub.Resume()
+				}
+
+				state = reestablished
+				exit = true
+			case aborted:
+				// todo: Terminate all subscribtions
+				c.cancelMonitor()
+				return
+			}
+		}
+		currentState <- state
+	}
+}
+
+// recreateSecureChannel create a new secure channel based on the configuration of the previous one
+func (c *Client) recreateSecureChannel(ctx context.Context) (*uasc.SecureChannel, *uacp.Conn, error) {
+	conn, err := uacp.Dial(ctx, c.endpointURL)
+	if err != nil {
+		return nil, nil, err
+	}
+	sechan, err := uasc.NewSecureChannel(c.endpointURL, conn, c.cfg)
+	if err != nil {
+		_ = conn.Close()
+		return nil, conn, err
+	}
+	return sechan, conn, nil
+}
+
+// repairSession repair the current session after being disconnected
+func (c *Client) repairSession() error {
+
+	s := c.Session()
+
+	if _, err := c.DetachSession(); err != nil {
+		return err
+	}
+
+	debug.Printf("Trying to reactivate existing session")
+
+	err := c.ActivateSession(s)
+	if err != nil {
+		return errors.Errorf("Session reactivation failed")
+	}
+	debug.Printf("Existing session reactivated trying to repair subscriptions")
+	if err := c.repairSubscriptions(); err != nil {
+		return err
+	}
+	debug.Printf("Subscriptions repaired")
+	return nil
+}
+
+// SubscriptionIDs gets a list of subscriptionIDs
+func (c *Client) SubscriptionIDs() []uint32 {
+	subscriptionIDs := []uint32{}
+	c.subMux.Lock()
+	defer c.subMux.Unlock()
+	for key := range c.subscriptions {
+		subscriptionIDs = append(subscriptionIDs, key)
+	}
+	return subscriptionIDs
+}
+
+// repairSubscriptions repair all the subscriptions of subscriptionIDs given,
+// if no subscriptionIDs repair all subscriptions
+func (c *Client) repairSubscriptions(subscriptionIDs ...uint32) error {
+
+	if subscriptionIDs == nil {
+		subscriptionIDs = c.SubscriptionIDs()
+	}
+
+	c.subMux.RLock()
+	defer c.subMux.RUnlock()
+
+	for _, subID := range subscriptionIDs {
+		sub, ok := c.subscriptions[subID]
+		if !ok {
+			return errors.Errorf("Invalid SubscriptionID")
+		}
+		if err := c.repairSubscription(sub); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) repairSubscription(sub *Subscription) error {
+	debug.Printf("RepairSubscription  for SubscriptionId %d", sub.SubscriptionID)
+	if status, err := c.subscriptionRepublish(sub); err != nil {
+		switch status {
```

Comment:
this looks fishy. the `status` is never evaluated, unless `err` is not nil

Reply:
Resolved in e0f222ba1a18d884479a5f5c433138578982c039